[SE-0031] Adjusting 'inout' Declarations for Type Decoration

dduan opened the pull request:
As the title indicates, this PR implements [SE-0031](https://github.com/apple/swift-evolution/blob/master/proposals/0031-adjusting-inout-declarations.md).
- [x] A warning is added for old `inout` position.
- [x] fixit for removing old inout position and inserting it in new position specified in `SE-0031`
- [x] when `inout` appears in both position, only the fixit for removal is issued
- [x] various places where inout declaration is printed
- [x] all tests are updated and passing.

dduan committed:
[Parser] update tests for 'inout' syntax adjustment

dduan committed:
[AST] update 'inout' print order in ASTPrinter

dduan committed:
[Parser] add test for 'inout' deprecation warning

chriseidhof commented on lib/Parse/ParsePattern.cpp:
```diff
@@ -186,9 +186,10 @@ Parser::parseParameterClause(SourceLoc &leftParenLoc,
       }
     }
-
+    SourceLoc depercatedInoutLoc;
```
There's a typo here: should be `deprecatedInoutLoc`

dduan replied:
@chriseidhof fixed. Thanks Chris!

dduan committed:
[Parser] adjust inout declaration for type decoration

This commit implements [SE-0031](//github.com/apple/swift-evolution/blob/master/proposals/0031-adjusting-inout-declarations.md).
When `inout` appears before parameter name, the parser issues a warning and
suggests the correct alternate location for it.

`inout` prefixing the paramter type is now valid.

lattner commented:
Totally awesome, thank you for working on this @dduan.  I have some specific comments which I'll comment on the diff itself.

lattner commented on include/swift/AST/DiagnosticsParse.def:
```diff
@@ -145,6 +145,9 @@ WARNING(lex_editor_placeholder_in_playground,none,
 NOTE(note_in_decl_extension,none,
      "in %select{declaration|extension}0 of %1", (bool, Identifier))
+WARNING(inout_as_attr_deprecated,none,
+        "'inout' as parameter attribute is deprecated, prefix type with it instead.",
```
This reads awkwardly, and diagnostic messages should not end with a period.  I'd suggest something like:

'inout' before a parameter name is deprecated, place it before the parameter type instead

lattner commented on lib/AST/ASTPrinter.cpp:
```diff
@@ -718,20 +720,18 @@ void PrintAST::printTypedPattern(const TypedPattern *TP) {
       Printer << "inout ";
     }
 
```
I think you can now remove  the special case "if (auto *IOT =" entirely.

lattner commented on lib/AST/ASTPrinter.cpp:
```diff
@@ -718,20 +720,18 @@ void PrintAST::printTypedPattern(const TypedPattern *TP) {
   if (auto *IOT = T->getAs<InOutType>()) {
     T = IOT->getObjectType();
     Printer << "inout ";
   }
```
Likewise, I think this "if" special case can be removed as well.

lattner commented on lib/AST/ASTPrinter.cpp:
```diff
@@ -1865,22 +1881,6 @@ void PrintAST::printOneParameter(const ParamDecl *param, bool Curried,
     }
   }
```
Similarly, both the InOutTypeRepr & the InOutType sides of this should be able to be dissolved now.  This probably allows this whole blob of code to go away (always a good sign :-)

dduan replied:
This one is different from the others. Skipping either branch causes a bunch of tests to fail.

lattner commented on lib/AST/ASTPrinter.cpp:
```diff
@@ -2831,16 +2831,16 @@ class TypePrinter : public TypeVisitor<TypePrinter> {
+        Printer << "inout ";
+        EltType = IOT->getObjectType();
+      }
       if (TD.isVararg()) {
```
Another one, the special case for inout should hopefully be possible to drop entirely.

lattner commented on lib/Parse/ParsePattern.cpp:
```diff
@@ -186,9 +186,10 @@ Parser::parseParameterClause(SourceLoc &leftParenLoc,
+      deprecatedInoutLoc = Tok.getLoc();
       param.LetVarInOutLoc = consumeToken();
       param.SpecifierKind = ParsedParameter::InOut;
     } else if (Tok.is(tok::kw_let)) {
```
I don't think you need the deprecatedInoutLoc variable at all here.  In the cases that use it, you can just use param.LetVarInOutLoc.  In the test for "deprecatedInoutLoc.isValid()", you can just check for "param.SpecifierKind == ParsedParameter::InOut".

lattner commented:
That's all I see.  The test suite changes all look great, please send a patch to update the standard library as well.  Thank you again for implementing this @dduan!

dduan commented:
Thanks for the review, Chris! I've addressed your comments to my best extent and pushed the changes up.

I have the patch for stdlib ready to go. Just waiting for this one to merge :)

lattner commented:
Awesome, thanks again!
